File: src/#965/solution.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val
  this.left = left === undefined ? null : left
  this.right = right === undefined ? null : right
}

var isUnivalTree = function (root) {
  let set = new Set()
  return helper(root)
  function helper(root) {
    if (root === null) {
      return true
    }
    if (set.size === 0) {
      set.add(root.val)
      return helper(root.left) && helper(root.right)
    } else if (set.size === 1 && set.has(root.val)) {
      return helper(root.left) && helper(root.right)
    }
    return false
  }
}

File: src/#965/solution.go
package solution

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isUnivalTree(root *TreeNode) bool {
	set := map[int]bool{}
	return helper(root, set)
}

func helper(root *TreeNode, set map[int]bool) bool {
	if root == nil {
		return true
	}

	if len(set) == 0 {
		set[root.Val] = true
		return helper(root.Left, set) && helper(root.Right, set)
	} else if _, ok := set[root.Val]; ok && len(set) == 1 {
		return helper(root.Left, set) && helper(root.Right, set)
	}
	return false
}
